Add postgres as a secondary DB connection.

DanVanAtta opened the pull request:
If the infrastructure is there, we would do parallel writes :)

Overview:
- Update Usercontroller so we can use either a derby connection, or a postgres. Turns out that is the only major difference between the two, just the connection that we pass in.
- Change how we do validation, instead of looking for null error string we have an explicit 'isValid' method, and a second method to get an error message.
- API updates for strong types when creating user, also provided a good home for user data validation.

Some next steps after this:
- update travis to set up DB schema
- update DbUserController test to also check a user controller that is using a postgres connection as primary.

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/login/LobbyLoginValidator.java:
```diff
@@ -176,31 +178,40 @@ private static String anonymousLogin(final Map<String, String> propertiesReadFro
       return null;
     } catch (final IllegalStateException ise) {
+      s_logger.severe("Error creating user: " + user + ", err: " + ise.getMessage());
+      ise.printStackTrace();
```
Is this intentional or a debugging session leftover?

DanVanAtta replied:
Intentional since java.util.Log does not do a stack trace. I know not best practice, but lacking a log4j/logback type logger that accepts an exception as an arg, I'm unaware of another way to get the stack trace.

ssoloff replied:
Is there an issue using [`Logger#log(Level, String, Throwable)`](https://docs.oracle.com/javase/8/docs/api/java/util/logging/Logger.html#log-java.util.logging.Level-java.lang.String-java.lang.Throwable-)?  For example:

```java
s_logger.log(Level.SEVERE, "Error creating user: " + user, ise);
```

ron-murhammer replied:
@DanVanAtta Bump this question.

ssoloff replied:
@DanVanAtta As @ron-murhammer pointed out, I forgot to dupe this question in my second round of comments.  While you're focused on this issue at the present, could you please review so it gets in this PR if appropriate?

DanVanAtta replied:
Good suggestion @ssoloff  - now updated 87642b6

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUser.java:
```diff
@@ -1,42 +1,98 @@
+
+  public static class UserName {
+    public final String value;
+    public UserName(String name) {
```
Checkstyle: "'CTOR_DEF' should be separated from previous statement."

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUser.java:
```diff
@@ -1,42 +1,98 @@
+      this.value = name;
+    }
+  }
+  public static class UserEmail {
```
Checkstyle: "'CLASS_DEF' should be separated from previous statement."

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUser.java:
```diff
@@ -1,42 +1,98 @@
+  }
+  public static class UserEmail {
+    public final String value;
+    public UserEmail(String email) {
```
Checkstyle: "'CTOR_DEF' should be separated from previous statement."

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUser.java:
```diff
@@ -1,42 +1,98 @@
+    return getValidationErrorMessage() == null;
   }
+  /**
+   * @return if this user is valid.
```
Looks like this may have been a copy-paste.  Please update to indicate correct return value description (including possible nullity).

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUser.java:
```diff
@@ -1,42 +1,98 @@
+  /**
+   * @return if this user is valid.
+   */
+  // TODO: test me
```
This method has 100% instruction coverage and 4 of 6 branches covered on line 57.  Not sure if the TODO is there to remind you to cover those remaining branches.  Otherwise, it can probably be removed.

DanVanAtta replied:
Thanks for spotting this. The TODO was not removed after I wrote DbUserTest, so I can remove it.

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUser.java:
```diff
@@ -1,42 +1,98 @@
+   * @return if this user is valid.
+   */
+  // TODO: test me
+  public String getValidationErrorMessage() {
```
Checkstyle: "'METHOD_DEF' should be separated from previous statement."

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUser.java:
```diff
@@ -1,42 +1,98 @@
+   * @return null if user name is valid, otherwise returns an error message.
+   */
+  public static String getInvalidUserNameReason(String userName) {
+    Preconditions.checkState(!isValidUserName(userName));
```
According to the Javadocs, this method will return `null` if the user name is valid, but won't this throw if the user name is valid?

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUserController.java:
```diff
@@ -18,39 +19,57 @@
+    this(
+        new DerbyUserController(
+            UserDaoPrimarySecondary.Role.PRIMARY,
+            () -> Database.getDerbyConnection()),
```
Nitpicking... these `Supplier`s can be replaced with method references, e.g. `Database::getDerbyConnection` and `Database::getPostgresConnection`.

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUserController.java:
```diff
@@ -18,39 +19,57 @@
 
-    return primaryDataSourceCandidates.get(0);
+    secondaryDaoSet = Arrays.stream(userDaos)
+        .filter(dao -> !dao.isPrimary())
```
Nitpicking... to be symmetric with line 55, this line can be replaced with `.filter(not(UserDaoPrimarySecondary::isPrimary))` where `not` is imported as

```java
import static games.strategy.util.PredicateUtils.not;
```

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUserController.java:
```diff
@@ -76,15 +95,14 @@ public void updateUser(final String name, final String email, final String hashe
     try {
-      // TODO : if there is a timeout on failure, this may need to be done on a background thread
-      implementations.forEach(dao -> dao.createUser(name, email, hashedPassword, admin));
+      new Thread(() -> secondaryDaoSet.forEach(dao -> dao.createUser(user, password)));
```
Missing a call to `Thread#start()`?

DanVanAtta replied:
eagle eye - good catch

ssoloff replied:
Not such a good eye...  I loaded this branch into Eclipse with static analysis enabled: `The allocated object is never used`. :smiley:

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUserController.java:
```diff
@@ -76,15 +95,14 @@ public void updateUser(final String name, final String email, final String hashe
     } catch (Exception e) {
       logger.warning(
-          "Secondary datasource failure, this is okay if we are still setting up a new datasource. " + e.getMessage());
+          "Secondary datasource failure, this is okay if we are still setting up a new datasource, error: " + e.getMessage());
```
Checkstyle: "Line is longer than 120 characters (found 126)."

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DerbyUserController.java:
```diff
@@ -69,71 +76,72 @@ public boolean doesUserExist(final String userName) {
+  public void updateUser(final DbUser user, final HashedPassword hashedPassword) {
+    Preconditions.checkArgument(user.isValid(), user.getValidationErrorMessage());
+
+    final Connection con = Database.getDerbyConnection();
```
Should this be using `connectionSupplier.get()` instead of `Database.getDerbyConnection()`?

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DerbyUserController.java:
```diff
@@ -69,71 +76,72 @@ public boolean doesUserExist(final String userName) {
       ps.close();
       con.commit();
     } catch (final SQLException sqle) {
+      // TODO: let's instead first check if the user exists, if we do that then we can just treat this as a normal error.
```
Checkstyle: "Line is longer than 120 characters (found 121)."

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/HashedPassword.java:
```diff
@@ -0,0 +1,20 @@
+package games.strategy.engine.lobby.server.userDB;
+
+import com.google.common.base.Strings;
+import games.strategy.util.MD5Crypt;
```
Checkstyle: "'games.strategy.util.MD5Crypt' should be separated from previous import group by one line."

ssoloff commented on src/test/java/games/strategy/engine/lobby/server/login/LobbyLoginValidatorTest.java:
```diff
@@ -13,6 +13,8 @@
 import games.strategy.engine.lobby.server.userDB.BadWordController;
+import games.strategy.engine.lobby.server.userDB.DbUser;
+import games.strategy.engine.lobby.server.userDB.HashedPassword;
 import games.strategy.engine.lobby.server.userDB.DbUserController;
```
Checkstyle: "Wrong lexicographical order for 'games.strategy.engine.lobby.server.userDB.DbUserController' import. Should be before 'games.strategy.engine.lobby.server.userDB.HashedPassword'."

ssoloff commented on src/test/java/games/strategy/engine/lobby/server/ModeratorControllerTest.java:
```diff
@@ -16,6 +16,8 @@
 
+import games.strategy.engine.lobby.server.userDB.DbUser;
+import games.strategy.engine.lobby.server.userDB.HashedPassword;
 import games.strategy.engine.lobby.server.userDB.DbUserController;
```
Checkstyle: "Wrong lexicographical order for 'games.strategy.engine.lobby.server.userDB.DbUserController' import. Should be before 'games.strategy.engine.lobby.server.userDB.HashedPassword'."

ssoloff commented on src/test/java/games/strategy/engine/lobby/server/userDB/DbUserTest.java:
```diff
@@ -0,0 +1,83 @@
+        invalidDbUser.isValid(), is(false));
+    assertThat(invalidDbUser.getValidationErrorMessage(), not(emptyString()));
+  }
+  private interface TestData {
```
Checkstyle: "'INTERFACE_DEF' should be separated from previous statement."

DanVanAtta commented:
One problem is new broken up classes are landing in an invalid package name. These are checkstyle issues which are hard to fix right now without renaming the 'userDB' package. I'm looking for two more random fixes to add to this branch to balance out the counts (For the modified files, unless something was missed violations would be due to package name)

DanVanAtta commented:
If we want, the commits from 7423968 and on can be lopped off into a new PR. We would need to increase the checkstyle count, as is this branch by finding some random fixes evens out the count since we added new classes to package "userDB" (+1 violation per new class..)

DanVanAtta commented:
https://github.com/triplea-game/triplea/pull/1797 opened to include the checkstyle fixes that were in  7423968 and forward. In the meantime will update the checkstyle counts since new violations cannot be readily fixed (due to existing package naming)

ssoloff commented:
> In the meantime I will update the checkstyle counts and increment for the new violations due to package naming.

:+1: I think bumping the Checkstyle thresholds is perfectly acceptable when, as you are in this PR, faced with issues that can't be easily changed due to compatibility or something similar.

ssoloff commented on src/integ_test/java/games/strategy/engine/lobby/server/userDB/DbUserControllerIntegrationTest.java:
```diff
@@ -15,90 +16,115 @@
+  private static int getUserCount(final Supplier<Connection> connection) throws Exception {
+    final Connection dbConnection = connection.get();
+    final String sql = "select count(*) from TA_USERS";
+    final ResultSet rs = connection.get().createStatement().executeQuery(sql);
```
Should `connection.get()` be `dbConnection`?  Otherwise, it's being orphaned.  Also, might want to consider using `dbConnection` in a try-with-resources block now that you are passing a `Supplier` to this method to avoid the explicit call to `close()` below.

ssoloff commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUser.java:
```diff
@@ -1,42 +1,103 @@
+   * @return null if user name is valid otherwise returns an error message String.
+   */
+  public static String getUserNameValidationErrorMessage(String userName) {
+    Preconditions.checkState(!isValidUserName(userName));
```
According to the Javadocs, this method will return `null` if the user name is valid, but won't this throw if the user name is valid?

(This comment was in the last review, but I think it got orphaned somehow.)

DanVanAtta commented:
Thanks for the review @ssoloff . I squashed everything together. @ron-murhammer anything you'd like to look at before this goes to master?

ron-murhammer commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUser.java:
```diff
@@ -1,42 +1,113 @@
-  private final Date joined;
+  public final String name;
+  public final String email;
+  public final boolean admin;
```
Would it not be better to have getters/setters here with private member variables?

DanVanAtta replied:
Setters would make these fields writable, that is not desirable.
We're following a pretty well trodden pattern for value objects here with an all arg constructor and public final member variables. Not having getters helps enforce the notion that the values are immutable. Ultimately getters would be okay, but they do not add anything.

ron-murhammer commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUser.java:
```diff
@@ -1,42 +1,113 @@
+    NOT_ADMIN, ADMIN
+  }
+
+  public static class UserName {
```
What's the thought here around the UserName and UserEmail classes with one field?

DanVanAtta replied:
It avoids the 'Stringly' typed values we had before. Instead of "new User(String,String,String)", we actually get some types "new User(User.UserName, User.UserEmail)"

A builder type pattern is something of an alternative, it would make these value types less useful.

For context, the value in these types has been shown while I was working on this PR. The HashPassword was split off to its own class when previously it was part of DbUser. That provided good flexibiilty and the updates to do that were minimal. We also had opportunity to consolidate some validation logic in that class, and it was suddenly much more testable as well.

ron-murhammer commented on src/main/java/games/strategy/engine/lobby/server/userDB/DbUser.java:
```diff
@@ -1,42 +1,113 @@
+    return userNameValidation(userName);
   }
+
+  private static String userNameValidation(String userName) {
```
Would this be better named as validateUserName?

ron-murhammer commented on src/main/java/games/strategy/engine/lobby/server/userDB/DerbyUserController.java:
```diff
@@ -156,18 +165,20 @@ public boolean login(final String userName, final String hashedPassword) {
 
   @Override
-  public DbUser getUser(final String userName) {
+  public DbUser getUserByName(final String userName) {
```
There are a lot of different variations of "username" (UserName, Name, username). It seems unclear whether it should be referred to as "user name" vs "username" vs "name".

DanVanAtta replied:
In part this is a reflection of the existing names. To another extent, it is 'user name' and called that pretty consistently. In honesty the 'user' prefix on everything is a potential concern and likely not very helpful and probably should be removed.
With that said, I'm not terribly bothered by the current variables. We could rename this method to `getUserByUserName`, but that seems really wordy to me.

DanVanAtta replied:
@ron-murhammer I think the latest two commits will address this concern a bit more - notably I ran across where we printed 'username' to external (to user). I updated that to be 'name' instead of 'username'.   Ideally IMO we'd drop 'username' in favor of 'name' when it is printed to users, and use 'userName' or 'name' internally. @ron-murhammer I think the latest two commits will address this concern a bit more - notably I ran across where we printed 'username' to external (to user). I updated that to be 'name' instead of 'username'.   Ideally IMO we'd drop 'username' in favor of 'name' when it is printed to users, and use 'userName' or 'name' internally.

ron-murhammer commented:
@DanVanAtta Comments added. Also looks like the travis file is conflicting.

DanVanAtta commented:
@ron-murhammer I added some test coverage to show a problem in validation and fixed that. With the test coverage then did some refactoring.
- validation of DbUser moved to the respective value objects, user name validation to userName, etc  (better cohesion, better consistency in validation code on DbUser).
- refactored away the internal String objects  in DbUser and replaced with the value objects. To avoid updating the types used by clients, I added getters to convert the value objects to string. Ideally we would continue using the stronger types until the last possible moment, but to keep the diff sane and avoid that scope creep, I've added the getters so clients can continue using `String` types. DbUser is still immutable, but no longer a POJO value object.
- The user name validation is where we printed 'username'. As mentioned, this was updated to 'name'.

ron-murhammer commented:
@DanVanAtta Need to resolve the merge conflict.

DanVanAtta commented:
https://github.com/triplea-game/triplea/pull/1792/commits/baa1359069bd38ecbc9db73d5601525400b48486 removes a leaked update. It is an interesting preview though I think of where I'm trying to get to verify the new DB connection.

ron-murhammer commented:
@DanVanAtta I still see a merge conflict with gradle.properties.

DanVanAtta commented:
Conflict is resolved now.

ssoloff commented on src/integ_test/java/games/strategy/engine/lobby/server/userDB/DbUserControllerIntegrationTest.java:
```diff
@@ -15,90 +16,118 @@
+    testCreate(Database::getDerbyConnection, new DbUserController());
+  }
+
+  private void testCreate(Supplier<Connection> connectionSupplier, DbUserController controller) throws Exception {
```
@DanVanAtta For your next PR...  I just noticed that the `create()` and `testCreate()` method names should probably be swapped.
